File: lib/sandbox/scenario.ml
type runner_platform =
  | Linux_x86_64
  | Linux_arm64
  | Windows_x86_64
  | Windows_arm64
  | Macos_x86_64
  | Macos_arm64

type t = {
  schema : string;
  digest : string;
  provider : Ir.provider;
  workflow_entrypoint : string;
  job : string;
  event : string;
  inputs : (string * string) list;
  matrix : (string * Json.t) list;
  variables : (string * string) list;
  runner_platform : runner_platform;
  secret_names : string list;
}

let runner_name = function
  | Linux_x86_64 -> "linux-x86_64"
  | Linux_arm64 -> "linux-arm64"
  | Windows_x86_64 -> "windows-x86_64"
  | Windows_arm64 -> "windows-arm64"
  | Macos_x86_64 -> "macos-x86_64"
  | Macos_arm64 -> "macos-arm64"

let runner_os = function
  | Linux_x86_64 | Linux_arm64 -> "linux"
  | Windows_x86_64 | Windows_arm64 -> "windows"
  | Macos_x86_64 | Macos_arm64 -> "macos"

let runner_of_string = function
  | "linux-x86_64" -> Some Linux_x86_64
  | "linux-arm64" -> Some Linux_arm64
  | "windows-x86_64" -> Some Windows_x86_64
  | "windows-arm64" -> Some Windows_arm64
  | "macos-x86_64" -> Some Macos_x86_64
  | "macos-arm64" -> Some Macos_arm64
  | _ -> None

let provider_of_string = function
  | "github" -> Some Ir.Github
  | "gitlab" -> Some Ir.Gitlab
  | "azure" -> Some Ir.Azure
  | "circleci" -> Some Ir.Circleci
  | _ -> None

let portable_name value =
  let first = function
    | 'A' .. 'Z' | 'a' .. 'z' | '_' -> true
    | _ -> false
  and rest = function
    | 'A' .. 'Z' | 'a' .. 'z' | '0' .. '9' | '_' | '-' | '.' -> true
    | _ -> false
  in
  String.length value > 0
  && first value.[0]
  && String.sub value 1 (String.length value - 1) |> String.for_all rest

let secret_name value =
  let first = function
    | 'A' .. 'Z' | 'a' .. 'z' | '_' -> true
    | _ -> false
  and rest = function
    | 'A' .. 'Z' | 'a' .. 'z' | '0' .. '9' | '_' -> true
    | _ -> false
  in
  String.length value > 0
  && first value.[0]
  && String.sub value 1 (String.length value - 1) |> String.for_all rest

let safe_entrypoint path =
  let path = Util.normalize_slashes path in
  path <> "" && Filename.is_relative path
  && (not (Util.starts_with ~prefix:"/" path))
  && (not (String.length path >= 2 && path.[1] = ':'))
  && Util.valid_utf8 path
  && path |> String.split_on_char '/'
     |> List.for_all (fun segment ->
         segment <> "" && segment <> "." && segment <> "..")

let canonical_pairs pairs =
  pairs
  |> List.map (fun (name, value) -> (name, value))
  |> List.sort (fun (left, _) (right, _) -> String.compare left right)

let string_object pairs =
  Json.Object (List.map (fun (name, value) -> (name, Json.String value)) pairs)

let unsigned_fields scenario =
  [
    ("event", Json.String scenario.event);
    ("inputs", string_object scenario.inputs);
    ("job", Json.String scenario.job);
    ("matrix", Json.Object scenario.matrix);
    ("provider", Json.String (Ir.provider_name scenario.provider));
    ("runner_platform", Json.String (runner_name scenario.runner_platform));
    ("schema", Json.String scenario.schema);
    ( "secret_names",
      Json.Array
        (List.map (fun value -> Json.String value) scenario.secret_names) );
    ("variables", string_object scenario.variables);
    ("workflow_entrypoint", Json.String scenario.workflow_entrypoint);
  ]

let unsigned_json scenario = Json.Object (unsigned_fields scenario)

let to_json scenario =
  Json.Object
    (("digest", Json.String scenario.digest) :: unsigned_fields scenario)

let to_canonical_json scenario = Json.to_string (to_json scenario) ^ "\n"

let unique_names context pairs =
  let names = List.map fst pairs in
  if List.length names = List.length (Util.deduplicate_strings names) then Ok ()
  else Error (context ^ " names must be unique")

let make ~provider ~workflow_entrypoint ~job ~event ~inputs ~matrix ~variables
    ~runner_platform ~secret_names =
  let workflow_entrypoint = Util.normalize_slashes workflow_entrypoint
  and inputs = canonical_pairs inputs
  and matrix = canonical_pairs matrix
  and variables = canonical_pairs variables
  and secret_names = List.sort String.compare secret_names in
  let open Util in
  let* () =
    if safe_entrypoint workflow_entrypoint then Ok ()
    else Error "scenario workflow_entrypoint must be a root-relative UTF-8 path"
  in
  let* () =
    if String.trim job <> "" then Ok ()
    else Error "scenario job must not be empty"
  in
  let* () =
    if String.trim event <> "" then Ok ()
    else Error "scenario event must not be empty"
  in
  let* () = unique_names "scenario input" inputs in
  let* () = unique_names "scenario matrix" matrix in
  let* () = unique_names "scenario variable" variables in
  let* () =
    if
      List.for_all (fun (name, _) -> portable_name name) (inputs @ variables)
      && List.for_all (fun (name, _) -> portable_name name) matrix
    then Ok ()
    else Error "scenario input, matrix, and variable names must be portable"
  in
  let* () =
    if
      List.for_all secret_name secret_names
      && List.length secret_names
         = List.length (Util.deduplicate_strings secret_names)
    then Ok ()
    else
      Error
        "scenario secret_names must be unique identifiers matching \
         [A-Za-z_][A-Za-z0-9_]*"
  in
  let provisional =
    {
      schema = "scenario-v1";
      digest = "";
      provider;
      workflow_entrypoint;
      job;
      event;
      inputs;
      matrix;
      variables;
      runner_platform;
      secret_names;
    }
  in
  let digest =
    "sha256:"
    ^ Sha256.digest_string (Json.to_string (unsigned_json provisional))
  in
  Ok { provisional with digest }

let parse_assignment value =
  match String.index_opt value '=' with
  | None -> Error ("expected NAME=VALUE: " ^ value)
  | Some index ->
      let name = String.sub value 0 index
      and contents =
        String.sub value (index + 1) (String.length value - index - 1)
      in
      if portable_name name then Ok (name, contents)
      else Error ("invalid portable assignment name: " ^ name)

let required name converter json =
  match Option.bind (Json.member name json) converter with
  | Some value -> Ok value
  | None -> Error ("scenario-v1 needs field " ^ name)

let string_pairs name json =
  let open Util in
  let* fields = required name Json.as_object json in
  let rec loop accumulator = function
    | [] -> Ok (List.rev accumulator)
    | (key, Json.String value) :: rest ->
        loop ((key, value) :: accumulator) rest
    | (key, _) :: _ -> Error (name ^ "." ^ key ^ " must be a string")
  in
  loop [] fields

let matrix_pairs json =
  let open Util in
  let* fields = required "matrix" Json.as_object json in
  if
    List.for_all
      (fun (_, value) ->
        match value with
        | Json.String _ | Json.Bool _ | Json.Int _ | Json.Int64 _ -> true
        | _ -> false)
      fields
  then Ok fields
  else
    Error "scenario matrix values must be scalar strings, booleans, or integers"

let string_list name json =
  let open Util in
  let* values = required name Json.as_array json in
  let rec loop accumulator = function
    | [] -> Ok (List.rev accumulator)
    | Json.String value :: rest -> loop (value :: accumulator) rest
    | _ -> Error (name ^ " must contain only strings")
  in
  loop [] values

let parse source =
  let open Util in
  let* json =
    match Json.parse source with
    | Ok value -> Ok value
    | Error error ->
        Error
          (Printf.sprintf "scenario JSON byte %d: %s" error.offset error.message)
  in
  let* _ =
    Json.exact_object ~context:"scenario-v1"
      ~allowed:
        [
          "digest";
          "event";
          "inputs";
          "job";
          "matrix";
          "provider";
          "runner_platform";
          "schema";
          "secret_names";
          "variables";
          "workflow_entrypoint";
        ]
      json
  in
  let* schema = required "schema" Json.as_string json in
  if schema <> "scenario-v1" then Error ("unsupported scenario schema " ^ schema)
  else
    let* supplied_digest = required "digest" Json.as_string json in
    let* provider_name = required "provider" Json.as_string json in
    let* provider =
      match provider_of_string provider_name with
      | Some value -> Ok value
      | None -> Error ("unknown scenario provider " ^ provider_name)
    in
    let* workflow_entrypoint =
      required "workflow_entrypoint" Json.as_string json
    in
    let* job = required "job" Json.as_string json in
    let* event = required "event" Json.as_string json in
    let* inputs = string_pairs "inputs" json in
    let* matrix = matrix_pairs json in
    let* variables = string_pairs "variables" json in
    let* runner_name = required "runner_platform" Json.as_string json in
    let* runner_platform =
      match runner_of_string runner_name with
      | Some value -> Ok value
      | None -> Error ("unknown runner platform " ^ runner_name)
    in
    let* secret_names = string_list "secret_names" json in
    let* scenario =
      make ~provider ~workflow_entrypoint ~job ~event ~inputs ~matrix ~variables
        ~runner_platform ~secret_names
    in
    if scenario.digest = supplied_digest then Ok scenario
    else Error "scenario digest mismatch"

File: crates/sandbox/src/scenario.rs
use std::collections::BTreeMap;
use workflow_verifier_domain::Provider;
use workflow_verifier_foundation::{JsonValue, PublicPath, content_digest};

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum RunnerPlatform {
    LinuxX86_64,
    LinuxArm64,
    WindowsX86_64,
    WindowsArm64,
    MacosX86_64,
    MacosArm64,
}

impl RunnerPlatform {
    #[must_use]
    pub fn name(self) -> &'static str {
        match self {
            Self::LinuxX86_64 => "linux-x86_64",
            Self::LinuxArm64 => "linux-arm64",
            Self::WindowsX86_64 => "windows-x86_64",
            Self::WindowsArm64 => "windows-arm64",
            Self::MacosX86_64 => "macos-x86_64",
            Self::MacosArm64 => "macos-arm64",
        }
    }

    #[must_use]
    pub fn os(self) -> &'static str {
        match self {
            Self::LinuxX86_64 | Self::LinuxArm64 => "linux",
            Self::WindowsX86_64 | Self::WindowsArm64 => "windows",
            Self::MacosX86_64 | Self::MacosArm64 => "macos",
        }
    }

    #[must_use]
    pub fn parse(value: &str) -> Option<Self> {
        Some(match value {
            "linux-x86_64" => Self::LinuxX86_64,
            "linux-arm64" => Self::LinuxArm64,
            "windows-x86_64" => Self::WindowsX86_64,
            "windows-arm64" => Self::WindowsArm64,
            "macos-x86_64" => Self::MacosX86_64,
            "macos-arm64" => Self::MacosArm64,
            _ => return None,
        })
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Scenario {
    pub digest: String,
    pub provider: Provider,
    pub workflow_entrypoint: String,
    pub job: String,
    pub event: String,
    pub inputs: BTreeMap<String, String>,
    pub matrix: BTreeMap<String, JsonValue>,
    pub variables: BTreeMap<String, String>,
    pub runner_platform: RunnerPlatform,
    pub secret_names: Vec<String>,
}

impl Scenario {
    /// Create a concrete scenario-v1 value with empty bindings.
    ///
    /// # Errors
    /// Rejects unsafe entrypoints and empty job or event selectors.
    pub fn new(
        provider: Provider,
        workflow_entrypoint: impl Into<String>,
        job: impl Into<String>,
        event: impl Into<String>,
        runner_platform: RunnerPlatform,
    ) -> Result<Self, String> {
        let mut scenario = Self {
            digest: String::new(),
            provider,
            workflow_entrypoint: workflow_entrypoint.into().replace('\\', "/"),
            job: job.into(),
            event: event.into(),
            inputs: BTreeMap::new(),
            matrix: BTreeMap::new(),
            variables: BTreeMap::new(),
            runner_platform,
            secret_names: Vec::new(),
        };
        scenario.validate()?;
        scenario.refresh_digest();
        Ok(scenario)
    }

    /// Add a portable string input and refresh the semantic digest.
    ///
    /// # Errors
    /// Rejects invalid or duplicate binding names.
    pub fn with_input(
        mut self,
        name: impl Into<String>,
        value: impl Into<String>,
    ) -> Result<Self, String> {
        let name = name.into();
        if !portable_name(&name) || self.inputs.insert(name, value.into()).is_some() {
            return Err("scenario input names must be unique and portable".to_owned());
        }
        self.refresh_digest();
        Ok(self)
    }

    /// Add a portable string variable and refresh the semantic digest.
    ///
    /// # Errors
    /// Rejects invalid or duplicate binding names.
    pub fn with_variable(
        mut self,
        name: impl Into<String>,
        value: impl Into<String>,
    ) -> Result<Self, String> {
        let name = name.into();
        if !portable_name(&name) || self.variables.insert(name, value.into()).is_some() {
            return Err("scenario variable names must be unique and portable".to_owned());
        }
        self.refresh_digest();
        Ok(self)
    }

    /// Add a scalar matrix value and refresh the semantic digest.
    ///
    /// # Errors
    /// Rejects invalid names, duplicates, and non-scalar values.
    pub fn with_matrix(
        mut self,
        name: impl Into<String>,
        value: JsonValue,
    ) -> Result<Self, String> {
        let name = name.into();
        if !matches!(
            value,
            JsonValue::String(_) | JsonValue::Boolean(_) | JsonValue::Integer(_)
        ) {
            return Err("scenario matrix values must be scalar".to_owned());
        }
        if !portable_name(&name) || self.matrix.insert(name, value).is_some() {
            return Err("scenario matrix names must be unique and portable".to_owned());
        }
        self.refresh_digest();
        Ok(self)
    }

    /// Declare a secret name without embedding its value.
    ///
    /// # Errors
    /// Rejects invalid or duplicate environment identifiers.
    pub fn with_secret(mut self, name: impl Into<String>) -> Result<Self, String> {
        let name = name.into();
        if !secret_name(&name) || self.secret_names.contains(&name) {
            return Err("scenario secret_names must be unique portable identifiers".to_owned());
        }
        self.secret_names.push(name);
        self.secret_names.sort();
        self.refresh_digest();
        Ok(self)
    }

    fn validate(&self) -> Result<(), String> {
        PublicPath::new(self.workflow_entrypoint.clone()).map_err(|_| {
            "scenario workflow_entrypoint must be a root-relative UTF-8 path".to_owned()
        })?;
        if self.job.trim().is_empty() {
            return Err("scenario job must not be empty".to_owned());
        }
        if self.event.trim().is_empty() {
            return Err("scenario event must not be empty".to_owned());
        }
        if !self
            .inputs
            .keys()
            .chain(self.variables.keys())
            .all(|name| portable_name(name))
            || !self.matrix.keys().all(|name| portable_name(name))
        {
            return Err("scenario input, matrix, and variable names must be portable".to_owned());
        }
        if !self.secret_names.iter().all(|name| secret_name(name)) {
            return Err("scenario secret_names must be portable identifiers".to_owned());
        }
        Ok(())
    }

    fn unsigned_fields(&self) -> BTreeMap<String, JsonValue> {
        BTreeMap::from([
            ("event".to_owned(), JsonValue::String(self.event.clone())),
            ("inputs".to_owned(), string_map(&self.inputs)),
            ("job".to_owned(), JsonValue::String(self.job.clone())),
            ("matrix".to_owned(), JsonValue::Object(self.matrix.clone())),
            (
                "provider".to_owned(),
                JsonValue::String(self.provider.name().to_owned()),
            ),
            (
                "runner_platform".to_owned(),
                JsonValue::String(self.runner_platform.name().to_owned()),
            ),
            (
                "schema".to_owned(),
                JsonValue::String("scenario-v1".to_owned()),
            ),
            (
                "secret_names".to_owned(),
                JsonValue::Array(
                    self.secret_names
                        .iter()
                        .cloned()
                        .map(JsonValue::String)
                        .collect(),
                ),
            ),
            ("variables".to_owned(), string_map(&self.variables)),
            (
                "workflow_entrypoint".to_owned(),
                JsonValue::String(self.workflow_entrypoint.clone()),
            ),
        ])
    }

    fn unsigned_json(&self) -> JsonValue {
        JsonValue::Object(self.unsigned_fields())
    }

    fn refresh_digest(&mut self) {
        self.digest = content_digest(self.unsigned_json().canonical());
    }

    #[must_use]
    pub fn verify_digest(&self) -> bool {
        self.digest == content_digest(self.unsigned_json().canonical())
    }

    #[must_use]
    pub fn to_canonical_json(&self) -> String {
        let mut fields = self.unsigned_fields();
        fields.insert("digest".to_owned(), JsonValue::String(self.digest.clone()));
        JsonValue::Object(fields).canonical_line()
    }

    /// Parse and authenticate strict scenario-v1 product JSON.
    ///
    /// # Errors
    /// Rejects malformed JSON, unknown fields, invalid values, and digest tampering.
    pub fn parse(source: &str) -> Result<Self, String> {
        let root = JsonValue::parse(source).map_err(|error| error.to_string())?;
        let fields = root.exact_object(
            "scenario-v1",
            &[
                "digest",
                "event",
                "inputs",
                "job",
                "matrix",
                "provider",
                "runner_platform",
                "schema",
                "secret_names",
                "variables",
                "workflow_entrypoint",
            ],
        )?;
        for name in [
            "digest",
            "event",
            "inputs",
            "job",
            "matrix",
            "provider",
            "runner_platform",
            "schema",
            "secret_names",
            "variables",
            "workflow_entrypoint",
        ] {
            if !fields.contains_key(name) {
                return Err(format!("scenario-v1 needs field {name}"));
            }
        }
        if string(fields, "schema")? != "scenario-v1" {
            return Err("unsupported scenario schema".to_owned());
        }
        let supplied_digest = string(fields, "digest")?.to_owned();
        let provider = match string(fields, "provider")? {
            "github" => Provider::Github,
            "gitlab" => Provider::Gitlab,
            "azure" => Provider::Azure,
            "circleci" => Provider::Circleci,
            value => return Err(format!("unknown scenario provider {value}")),
        };
        let runner_name = string(fields, "runner_platform")?;
        let runner_platform = RunnerPlatform::parse(runner_name)
            .ok_or_else(|| format!("unknown runner platform {runner_name}"))?;
        let mut scenario = Self::new(
            provider,
            string(fields, "workflow_entrypoint")?,
            string(fields, "job")?,
            string(fields, "event")?,
            runner_platform,
        )?;
        scenario.inputs = parse_string_map(fields, "inputs")?;
        scenario.variables = parse_string_map(fields, "variables")?;
        scenario.matrix = parse_matrix(fields)?;
        scenario.secret_names = fields
            .get("secret_names")
            .and_then(JsonValue::as_array)
            .ok_or_else(|| "secret_names must be an array".to_owned())?
            .iter()
            .map(|value| {
                value
                    .as_str()
                    .map(str::to_owned)
                    .ok_or_else(|| "secret_names must contain strings".to_owned())
            })
            .collect::<Result<Vec<_>, _>>()?;
        let mut unique = scenario.secret_names.clone();
        unique.sort();
        unique.dedup();
        if unique.len() != scenario.secret_names.len() {
            return Err("scenario secret_names must be unique".to_owned());
        }
        scenario.secret_names.sort();
        scenario.validate()?;
        scenario.refresh_digest();
        if scenario.digest != supplied_digest {
            return Err("scenario digest mismatch".to_owned());
        }
        Ok(scenario)
    }
}

fn portable_name(value: &str) -> bool {
    let mut bytes = value.bytes();
    bytes
        .next()
        .is_some_and(|byte| byte.is_ascii_alphabetic() || byte == b'_')
        && bytes.all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'-' | b'.'))
}

fn secret_name(value: &str) -> bool {
    let mut bytes = value.bytes();
    bytes
        .next()
        .is_some_and(|byte| byte.is_ascii_alphabetic() || byte == b'_')
        && bytes.all(|byte| byte.is_ascii_alphanumeric() || byte == b'_')
}

fn string_map(values: &BTreeMap<String, String>) -> JsonValue {
    JsonValue::Object(
        values
            .iter()
            .map(|(name, value)| (name.clone(), JsonValue::String(value.clone())))
            .collect(),
    )
}

fn string<'a>(fields: &'a BTreeMap<String, JsonValue>, name: &str) -> Result<&'a str, String> {
    fields
        .get(name)
        .and_then(JsonValue::as_str)
        .ok_or_else(|| format!("scenario-v1 needs string field {name}"))
}

fn parse_string_map(
    fields: &BTreeMap<String, JsonValue>,
    name: &str,
) -> Result<BTreeMap<String, String>, String> {
    let JsonValue::Object(values) = fields
        .get(name)
        .ok_or_else(|| format!("scenario-v1 needs field {name}"))?
    else {
        return Err(format!("scenario {name} must be an object"));
    };
    values
        .iter()
        .map(|(key, value)| {
            value
                .as_str()
                .map(|value| (key.clone(), value.to_owned()))
                .ok_or_else(|| format!("scenario {name}.{key} must be a string"))
        })
        .collect()
}

fn parse_matrix(
    fields: &BTreeMap<String, JsonValue>,
) -> Result<BTreeMap<String, JsonValue>, String> {
    let JsonValue::Object(values) = fields
        .get("matrix")
        .ok_or_else(|| "scenario-v1 needs field matrix".to_owned())?
    else {
        return Err("scenario matrix must be an object".to_owned());
    };
    if values.values().all(|value| {
        matches!(
            value,
            JsonValue::String(_) | JsonValue::Boolean(_) | JsonValue::Integer(_)
        )
    }) {
        Ok(values.clone())
    } else {
        Err("scenario matrix values must be scalar strings, booleans, or integers".to_owned())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scenario() -> Scenario {
        Scenario::new(
            Provider::Github,
            ".github/workflows/ci.yml",
            "build",
            "push",
            RunnerPlatform::LinuxX86_64,
        )
        .unwrap()
    }

    #[test]
    fn validation_checks_binding_and_matrix_names_independently() {
        let mut invalid_input = scenario();
        invalid_input
            .inputs
            .insert("9input".to_owned(), "value".to_owned());
        assert_eq!(
            invalid_input.validate().unwrap_err(),
            "scenario input, matrix, and variable names must be portable"
        );

        let mut invalid_matrix = scenario();
        invalid_matrix
            .matrix
            .insert("9matrix".to_owned(), JsonValue::Boolean(true));
        assert_eq!(
            invalid_matrix.validate().unwrap_err(),
            "scenario input, matrix, and variable names must be portable"
        );
    }
}
